### ISSUE TITLE
Wiped out my previous config and have gone all in :) ... slight issue with 'attrunique'

### ISSUE THREAD

**Issue by mannp** [oddlama/kanidm-provision#3]
Hi @oddlama 

Hate asking, but having an issue that keeps occurring;

`kanidm-provision 1.0.1`

syncing groups
Syncing persons
Syncing oauth2 resource servers
    Creating /v1/oauth2/_basic/forgejo-local
```
Error: 
   0: body: {"plugin":{"attrunique":"duplicate value detected"}}
   1: HTTP status server error (500 Internal Server Error) for url (https://idm.domain.com/v1/oauth2/_basic)
```

```
Location:
   src/client.rs:47
```

I've since deleted the db and started again in case there were left over issues, but it happened again on a clean db.

Is there any way of knowing what is duplicated at the moment?

Cheers

```
    "systems": {
        "oauth2": {
            "headscale-local": {
                "allowInsecureClientDisablePkce": true,
                "basicSecretFile": null,
                "claimMaps": {},
                "displayName": "Headscale",
                "originLanding": "https://headscale.domain.com/oidc/callback",
                "originUrl": "https://headscale.domain.com/",
                "preferShortUsername": false,
                "present": true,
                "removeOrphanedClaimMaps": true,
                "scopeMaps": {
                    "headscale-access": [
                        "openid",
                        "email",
                        "profile"
                    ]
                },
                "supplementaryScopeMaps": {}
            },
            "forgejo-local": {
                "allowInsecureClientDisablePkce": true,
                "basicSecretFile": null,
                "claimMaps": {},
                "displayName": "Forgejo",
                "originLanding": "https://git.domain.com/user/oauth2/idm/callback/",
                "originUrl": "https://git.domain.com/",
                "preferShortUsername": true,
                "present": true,
                "removeOrphanedClaimMaps": true,
                "scopeMaps": {
                    "forgejo-access": [
                        "openid",
                        "email",
                        "profile"
                    ]
                },
                "supplementaryScopeMaps": {}
            }
        }
    }
```

```
  ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ BACKTRACE ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━
                                ⋮ 3 frames hidden ⋮                               
   4: <reqwest::blocking::response::Response as kanidm_provision::client::ResponseExt>::detailed_error_for_status::h760950d06ef5d630
      at <unknown source file>:<unknown line>
   5: kanidm_provision::client::KanidmClient::create_entity::hf4e04c5fed1f1eda
      at <unknown source file>:<unknown line>
   6: kanidm_provision::sync_oauth2s::ha6a2f9569a4ba69c
      at <unknown source file>:<unknown line>
   7: kanidm_provision::main::h6bdb0d2e4e882a4f
      at <unknown source file>:<unknown line>
   8: std::sys_common::backtrace::__rust_begin_short_backtrace::hd802b91d3bdf6f4c
      at <unknown source file>:<unknown line>
   9: std::rt::lang_start::{{closure}}::h3f4829b68332ddb9
      at <unknown source file>:<unknown line>
  10: std::panicking::try::hbc3e77d512204ae2
      at <unknown source file>:<unknown line>
  11: std::rt::lang_start_internal::hf04efd76000e5213
      at <unknown source file>:<unknown line>
  12: main<unknown>
      at <unknown source file>:<unknown line>
  13: __libc_start_call_main<unknown>
      at <unknown source file>:<unknown line>
  14: __libc_start_main@@GLIBC_2.34<unknown>
      at <unknown source file>:<unknown line>
  15: _start<unknown>
      at <unknown source file>:<unknown line>

```

**Comment by oddlama** [comment]
Do the numerals work in kanidm itself?

**Comment by mannp** [comment]
> Do the numerals work in kanidm itself?

For me yes, I was able to add the form `forgejo-local-msYxfofsdfsdfloYGHF97r6BBG` to kanidm without an issue, but when I took that naming over to here I started having problems.

After a while of trial and error I am now using `forgejo-word1-word2-word3` format and your tools is working perfectly.

I am not precious either way :) but might be worth a note.

**Comment by oddlama** [comment]
Definitely, thanks for documenting this. Maybe I'll have time at some point to add support for this. :)

**Comment by amnnku** [comment]
> Definitely, thanks for documenting this. Maybe I'll have time at some point to add support for this. :)

One thing I wasn't clear on which you could perhaps clear up.

Is there a 'just wipe things out and start a fresh' meethod, as when having problems, I went from one issue to another with left over items in the kanidm db.

**Comment by mannp** [comment]
One thing I wasn't clear on which you could perhaps clear up.

Is there a 'just wipe things out and start a fresh' meethod, as when having problems, I went from one issue to another with left over items in the kanidm db.

**Comment by oddlama** [comment]
Not directly, at least not from this tool. There is a "wipe everything method" in kanidm of course, just delete the service's state folder in it's entirety.

There just are a lot of different entities in kanidm and I don't want the tool to interfere with anything that exists in kanidm (or might exist in the future). If we cannot modify it, we cannot modify it accidentally or due to misconfiguration. Therefore, the tool will very strictly only touch things that it is directly responsible for, meaning that it had been mentioned in the config at some point. Any errors that arise from preexisting configuration are telling users that there might be stuff that conflicts with provisioned entities (which would otherwise be overwritten, and that would be undesirable).

Unfortunately, this also means that there probably is no easy way to support "migration" of existing entities to the provisioning system. Starting off clean is probably the most robust way. Although I'm curious, what things exactly were leftover by kanidm that caused errors for you? Personally I was always able to delete an entity in kanidm and then start provisioning it in my tests without any problems.

**Comment by mannp** [comment]
> Not directly, at least not from this tool. There is a "wipe everything method" in kanidm of course, just delete the service's state folder in it's entirety.

Yes I did that and tried to provision here, but the tool complained from the off. It was hard to tell at that point though, as I was using the problematic naming, which seemed to give me different errors which were hard to follow.

> There just are a lot of different entities in kanidm and I don't want the tool to interfere with anything that exists in kanidm (or might exist in the future). If we cannot modify it, we cannot modify it accidentally or due to misconfiguration. Therefore, the tool will very strictly only touch things that it is directly responsible for, meaning that it had been mentioned in the config at some point. Any errors that arise from preexisting configuration are telling users that there might be stuff that conflicts with provisioned entities (which would otherwise be overwritten, and that would be undesirable).

I think I may have misunderstood the meaning of 'orphan removal' as I interpreted that to mean, it would delete anything else in kanidm that wasn't provisioned with this tool.

But after using it, it might mean more that it will only remove the things that it was managing from the start and nothing else already existing in kanidm.

I also seemed to misunderstand also the meaning of the to me cryptic error `"attrunique":"duplicate value detected"`, which I thought was relating to duplicate entries in my state.json, but now you have explained a little more, it might mean duplicates between the state.json and the existing kanidm state.

> Although I'm curious, what things exactly were leftover by kanidm that caused errors for you?

At the time It was more that I had added multiple items through the kanidm manually, and expected your tool to wipe and overwrite with the contents of state.json.

Now I understand better, I appreciate the design, just didn't pick it up from the start :)

**Comment by oddlama** [comment]
> I think I may have misunderstood the meaning of 'orphan removal' as I interpreted that to mean, it would delete anything else in kanidm that wasn't provisioned with this tool.
> But after using it, it might mean more that it will only remove the things that it was managing from the start and nothing else already existing in kanidm.

Yeah, I should probably clarify this in the readme. 

> I also seemed to misunderstand also the meaning of the to me cryptic error `"attrunique":"duplicate value detected"`, which I thought was relating to duplicate entries in my state.json, but now you have explained a little more, it might mean duplicates between the state.json and the existing kanidm state.

I see, do you think it would have been less misleading if the error message clearly showed that this is the response that kanidm returned? If the kanidm API reports an error, the tool just prints whatever it received from kanidm. This already is the maximum amount of information we have if something goes wrong (unfortunately the API responses are often lacking context, but we can't assume what went wrong)

**Comment by mannp** [comment]
> Yeah, I should probably clarify this in the readme.

Certainly might help others :)

> I see, do you think it would have been less misleading if the error message clearly showed that this is the response that kanidm returned? If the kanidm API reports an error, the tool just prints whatever it received from kanidm. This already is the maximum amount of information we have if something goes wrong (unfortunately the API responses are often lacking context, but we can't assume what went wrong)

I absolutely do, and think it's better for your app to make that clear. The error was meaningless to me, next to useless.

Making it clear any output is from kanidm, would have been helpful to me at least, so maybe for others too :)

Thanks again for the tool, once working, it saved me a lot of manual labour :)
